### PR TITLE
Close #LGVISIUM-57: Downgraded the PyMuPDF version and updated the README

### PR DESCRIPTION
**Pull request by dcleres**
Some of our PDF documents have performance issues with newer versions of PyMuPDF. This change in the code makes it possible to use an older version of the library, which does not present these performance issues.

### CODE REVIEW

**Comment by dcleres** [comment]
Screenshot 2024-08-22 at 17 07 44

When using the 1.24.1 version of the code, the change of speed between drawing on the Page or on the Shape (new version) is not large. See 2.5 min to 2.4 min. However, this makes the code compatible with the more recent versions of the PyMuPDF and was the way suggested by one of the developers. (See https://github.com/pymupdf/PyMuPDF/issues/3800)

I added the functionality to log the runtime per document. @stijnvermeeren-swisstopo if you think that it is not necessary we can remove it. However, I do believe that every time we will try to improve the performance we will have to rewrite this code.

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
Thanks a lot for reaching out to the PyMuPDF devs and for identifying the cause of the performance change!

I managed to further improve the performance by reducing the number of shape.commit() calls. First of all, the one in draw_layer was redundant anyway. Then I managed to unify the remaining five calls per page to a single one.

With this, I would also be happy to upgrade again to the latest version of PyMuPDF, as the performance advantages of the old version no longer seem to outweigh the benefits of being able to use the latest and greatest library version ;).

I also refactored the creation of the "timing" dataframes to work with a list of dicts instead of a dict of lists. For me, the new way is cleaner and slightly simpler.

File: src/stratigraphy/main.py [hunk 1]
```diff
@@ -226,19 +237,28 @@ def start_pipeline(
...
         )  # mlflow.log_artifact expects a file
+        processing_time_df.to_csv(temp_directory / "processing_times.csv", index=False)
+        total_runtime = time.time() - start_time_pipeline
+        logger.info("Total runtime: %s", total_runtime)
 
         if mlflow_tracking:
             mlflow.log_metrics(metrics)
             mlflow.log_artifact(temp_directory / "document_level_metrics.csv")
+            mlflow.log_artifact(temp_directory / "processing_times.csv")
+            mlflow.log_metric("total_runtime", round(total_runtime, 2))
```
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/main.py:257]: Yes, this is true. I left it as I saw the value of tracking the processing time for each single file, which is currently not done by mlflow (correct me if I am wrong) as this is how you detected the drop in performance in the first place. I was thinking that at a later time, we could have a specific test that makes sure that no file takes longer than, let's say, 10 seconds to be annotated. But if this is not necessary, then I am happy to remove this artifact from the code.

**Comment by dcleres** [comment]
> I managed to further improve the performance by reducing the number of shape.commit() calls. First of all, the one in draw_layer was redundant anyway. Then I managed to unify the remaining five calls per page to a single one.

That is great! 

I will clean up the code and make push the new changes this afternoon.

File: src/stratigraphy/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/main.py:257]: @stijnvermeeren-swisstopo let me know what you think.

**Comment by dcleres** [comment]
Screenshot 2024-08-26 at 15 03 29

I tried the code with the latest changes (on top), and we are pretty much on par with the performance we had with the older version (on the bottom). Super cool!

File: src/stratigraphy/main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcleres) [reply src/stratigraphy/main.py:257]: Ah okay. Then I will remove it.

**Comment by dcleres** [comment]
@stijnvermeeren-swisstopo the code should not log single file timing information anymore. Are we good to merge?

**Pull request merged**